### PR TITLE
status wrong for directories when using vfs and nextcloud desktop 3.2.0

### PR DESCRIPTION
**Pull request by mgallien**
#3082

### CODE REVIEW

File: src/libsync/vfs/cfapi/cfapiwrapper.cpp [hunk 1]
```diff
@@ -58,6 +58,12 @@ void cfApiSendTransferInfo(const CF_CONNECTION_KEY &connectionKey, const CF_TRAN
         qCCritical(lcCfApiWrapper) << "Couldn't send transfer info" << QString::number(transferKey.QuadPart, 16) << ":" << cfExecuteresult << QString::fromWCharArray(_com_error(cfExecuteresult).ErrorMessage());
     }
 
+    // Once the download is finished, there is no need to do a last update of the progress status
+    if (offset + currentBlockLength == totalLength)
+    {
```
Comment (er-vin) [inline src/libsync/vfs/cfapi/cfapiwrapper.cpp:63]: style nitpick: brace should be on previous line

One day we'll have clang-format-diff checks on the CI side and I'll put my OCD to rest. ;-)

File: src/libsync/vfs/cfapi/cfapiwrapper.cpp [hunk 2]
```diff
@@ -694,7 +700,7 @@ OCC::Result<void, QString> OCC::CfApiWrapper::convertToPlaceholder(const FileHan
 
     const auto fileIdentity = QString::fromUtf8(fileId).toStdWString();
     const auto fileIdentitySize = (fileIdentity.length() + 1) * sizeof(wchar_t);
-    const qint64 result = CfConvertToPlaceholder(handle.get(), fileIdentity.data(), sizeToDWORD(fileIdentitySize), CF_CONVERT_FLAG_NONE, nullptr, nullptr);
+    const qint64 result = CfConvertToPlaceholder(handle.get(), fileIdentity.data(), sizeToDWORD(fileIdentitySize), CF_CONVERT_FLAG_MARK_IN_SYNC, nullptr, nullptr);
```
Comment (er-vin) [inline src/libsync/vfs/cfapi/cfapiwrapper.cpp:703]: That fills a bit heavy handed to me since it'll mark it sync indiscriminately all placeholders which I guess it might not always be the case... or my memory of that API gets fuzzier now. Isn't it something we'd rather deal with from `VfsCfApi::fileStatusChanged`?

Currently its implementation is empty because in my tests I couldn't find anything really wrong with having CF_CONVERT_FLAG_NONE and nothing more. But apparently I was looking at the wrong place.

**Review by allexzander — APPROVED** [review]
LGTM. And, agree about Kevin's comment.

File: src/libsync/vfs/cfapi/cfapiwrapper.cpp (hunk 2 above)
Comment (mgallien) [reply src/libsync/vfs/cfapi/cfapiwrapper.cpp:703]: As far as I understand, HydrationJob are implied when some software (via the user) wants to access a file that has not been hydrated.

If the file is being downloaded because it has been pinned, we do not use the HydrationJob but we always go through `CfConvertToPlaceholder`.

I did *some* tests with the change and could not notice issues. I can try harder to find issues.

File: src/libsync/vfs/cfapi/cfapiwrapper.cpp (hunk 1 above)
Comment (mgallien) [reply src/libsync/vfs/cfapi/cfapiwrapper.cpp:63]: Thanks.

I am not sure if the comment I have added is really needed here. What do you think ?

File: src/libsync/vfs/cfapi/cfapiwrapper.cpp (hunk 2 above)
Comment (er-vin) [reply src/libsync/vfs/cfapi/cfapiwrapper.cpp:702]: > As far as I understand, HydrationJob are used when some software (via the user) wants to access a file that has not been hydrated.

Right and so this should handle the flag change for us I guess?

> If the file is being downloaded because it has been pinned, we do not use the HydrationJob but we always go through `CfConvertToPlaceholder`.

Indeed. Is it all files or only for files which are unavailable? I admit I don't quite remember... That's also where I'm not sure if CF_CONVERT_FLAG_MARK_IN_SYNC is really appropriate in all cases. Because there are the transient states like "it needs a sync" and "it's syncing". I wonder how that plays along with those.

As you can tell I'm mostly brain dumping here. :-)

> I did _some_ tests with the change and could not notice issues. I can try harder to find issues.

I'd look around transient states mostly (makes them harder to test though). Might not be a problem in practice.

File: src/libsync/vfs/cfapi/cfapiwrapper.cpp (hunk 1 above)
Comment (er-vin) [reply src/libsync/vfs/cfapi/cfapiwrapper.cpp]: Well, you felt the need for a comment so I guess something is needed for readability. Admittedly, in situations like this, I'm more a proponent of well named variables before comments (I keep comments as last resort). In the case at hands it'd give:
```cpp
const auto isDownloadFinished = ((offset + currentBlockLength) == totalLength);
if (isDownloadFinished) {
```

Note: not sure I'd always prescribe using parentheses on the whole expression like I did here, just felt more readable to me for this one somehow (probably the number of = on the line and the distance between = and == pushed me there, no rule to make out of that).

File: src/libsync/vfs/cfapi/cfapiwrapper.cpp (hunk 2 above)
Comment (allexzander) [reply src/libsync/vfs/cfapi/cfapiwrapper.cpp:702]: @er-vin Documentation for this API feels incomplete, really. From my tests on Matthiue's branch, this status icon's state behaves much better now.

**Comment by er-vin** [comment]
> @er-vin Documentation for this API feels incomplete, really.

Yes... We're badly spoiled with the Qt API documentation. ;-)

> From my tests on Matthiue's branch, this status icon's state behaves much
> better now.

I don't doubt it, just pushing to make sure there's no annoying corner case 
lurking. They're easier to fix when the change is introduced than later on.

<insert plot about cost of bug fix over age of bug here> :-)

**Comment by allexzander** [comment]
@er-vin 
> I don't doubt it, just pushing to make sure there's no annoying corner case lurking. They're easier to fix when the change is introduced than later on. <insert plot about cost of bug fix over age of bug here> :-)

This is also true, but, with this new CfAPI thing, I am afraid, this is not the last issue. We'd still need some feedback from users to see how things work out in this case of, rather poorly documented API. Not sure how to check the correctness of this state right now. But, surely, this part is not risky at all, since it's just the front-end part that does not affect the hydration/dehadration. And this is also the reason why it's hard to validate it. It's because this front-end part is in Windows' internals completely :(

**Comment by er-vin** [comment]
> @er-vin
> 
> > I don't doubt it, just pushing to make sure there's no annoying corner case lurking. They're easier to fix when the change is introduced than later on.  :-)
> 
> This is also true, but, with this new CfAPI thing, I am afraid, this is not the last issue. We'd still need some feedback from users to see how things work out in this case of, rather poorly documented API. Not sure how to check the correctness of this state right now.

Unfortunately I have to agree. And that's why I like to limit platform specific code as much as possible, it often ends up like this otherwise. :-)

> But, surely, this part is not risky at all, since it's just the front-end part that does not affect the hydration/dehadration.

Which in itself is "interesting" isn't it? The API pushes to claim it's synced while data is not there... :disappointed:

> And this is also the reason why it's hard to validate it. It's because this front-end part is in Windows' internals completely :(

Yes, there *seem* to be some lower level APIs which allow to drill into those states somehow... but making automated tests leveraging them is gruesome.

File: src/libsync/vfs/cfapi/cfapiwrapper.cpp (hunk 2 above)
Comment (mgallien) [reply src/libsync/vfs/cfapi/cfapiwrapper.cpp:702]: > > As far as I understand, HydrationJob are used when some software (via the user) wants to access a file that has not been hydrated.
> 
> Right and so this should handle the flag change for us I guess?

And this part is showing the correct status. I had just needed to avoid having the progress bar reappear when the download is completed.

> 
> > If the file is being downloaded because it has been pinned, we do not use the HydrationJob but we always go through `CfConvertToPlaceholder`.
> 
> Indeed. Is it all files or only for files which are unavailable? I admit I don't quite remember... That's also where I'm not sure if CF_CONVERT_FLAG_MARK_IN_SYNC is really appropriate in all cases. Because there are the transient states like "it needs a sync" and "it's syncing". I wonder how that plays along with those.

If I understand what happen, we have:

1) the user modifies the pinning state of some files that were cloud only (not hydrated)
2) our directory watcher detects that and we see that files have to be downloaded
3) we convert the plain old files into placeholders via `CfConvertToPlaceholder`

I suspect that indeed I have not tested enough what happen if a files is half downloaded. I suppose that we already have quite some code handling that given we are just using the pre-existing code path to do that. This is just after completeion that we have special handling of CfApi.

> 
> As you can tell I'm mostly brain dumping here. :-)
> 
> > I did _some_ tests with the change and could not notice issues. I can try harder to find issues.
> 
> I'd look around transient states mostly (makes them harder to test though). Might not be a problem in practice.

I will. Thanks for the suggestion

File: src/libsync/vfs/cfapi/cfapiwrapper.cpp (hunk 1 above)
Comment (mgallien) [reply src/libsync/vfs/cfapi/cfapiwrapper.cpp]: This is indeed what I should have done at first.

I noticed that my comment was a sign of code not being good enough when reading back what I had written.

File: src/libsync/vfs/cfapi/cfapiwrapper.cpp (hunk 2 above)
Comment (mgallien) [reply src/libsync/vfs/cfapi/cfapiwrapper.cpp:702]: I did more tests and could only get an error when trying implicit hydration at the same time files where being downloaded by pinning.
At all time, it looks like file status were correct.

Unrelated note: When trying to open the same file twice, I got two download and the second one terminated with an error.

Comment (allexzander) [reply src/libsync/vfs/cfapi/cfapiwrapper.cpp:702]: @mgallien Would be good to detect "this file is already being hydrated" and then - early-out from the second hydration attempt.

**Comment by mgallien** [comment]
So, when a file is being downloaded because it is being pinned and the user is also trying to open it (with implicit hydration being requested), we will get an abort on the file sync job. There is even a TODO comment to fix that. This code is still the same on the original repository where this is code is coming from. So something we should eventually fix but not now.

File: src/libsync/vfs/cfapi/cfapiwrapper.cpp (hunk 2 above)
Comment (mgallien) [reply src/libsync/vfs/cfapi/cfapiwrapper.cpp:702]: Implicit hydration of the same file two times is not triggering any error and is behaving like OneDrive. I believe we are OK there.

**Review by er-vin — COMMENTED** [review]
LGTM
